{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# popDMS\n",
    "\n",
    "**A population genetics-based method for inferring mutation effects from deep mutational scanning (DMS) data**\n",
    "\n",
    "popDMS takes input codon counts in [dms_tools format](https://jbloomlab.github.io/dms_tools/fileformats.html#deep-mutational-scanning-counts-file) or sequence counts in [MAVE-HGVS format](https://www.mavedb.org/docs/mavehgvs/) and infers selection coefficients for each amino acid at each site in the data. Roughly, selection coefficients can be interpreted as the fractional advantage (if positive) or disadvantage (if negative) for each variant in passing through selection in the experiment.\n",
    "\n",
    "Below, we include a basic guide for how to use the main functions of popDMS to 1) compute variant frequencies and correlations from count data, and 2) to infer selection coefficients from the computed frequencies. For more details and applications to example data sets, see [the repository](https://github.com/bartonlab/paper-DMS-inference) for the paper describing popDMS. Usage is slightly different when using codon counts and when using sequence counts. We also show how to visualize the inferred selection coefficients with a heatmap."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example using codon counts\n",
    "\n",
    "Let's assume that we have a DMS data set that we refer to as `example`. In this data set, there were two experimental replicates. Codon counts before selection were saved in two files, `example_mutDNA_1.csv` and `example_mutDNA_2.csv` for replicates 1 and 2, respectively. Codon counts after selection, which here we will assume is after one generation of replication, are stored in files `example_mutvirus_1.csv` and `example_mutvirus_2.csv`. The codon counts should be in [dms_tools format](https://jbloomlab.github.io/dms_tools/fileformats.html#deep-mutational-scanning-counts-file) and should be readable by `pandas`. \n",
    "\n",
    "The code block below would analyze this data set and store the inferred selection coefficients in a compressed CSV file `example_selection_coefficients.csv.gz`. This file can either be analyzed by `pandas` or similar packages while compressed, or it can be uncompressed to be viewed directly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import popDMS\n",
    "\n",
    "\n",
    "#___ popDMS parameters\n",
    "\n",
    "## `corr_cutoff_pct` sets the maximum fraction of drop in correlation per log10(gamma) for setting optimal regularization strength \n",
    "## 0.1 is a good default value\n",
    "corr_cutoff_pct = 0.1  \n",
    "\n",
    "\n",
    "#___ data parameters\n",
    "\n",
    "## `name` gives the name of the data set, which will be prepended to the files that popDMS saves\n",
    "name = 'example'\n",
    "\n",
    "## `n_replicates` specifies the number of experimental replicates for this data set\n",
    "n_replicates = 2\n",
    "\n",
    "## `codon_counts_files` specifies the path to the codon counts files\n",
    "## the order of the files is arbitrary, but subsequent variables (`replicates`, `times`) will assume the same order\n",
    "codon_counts_files = ['example_mutDNA_1.csv',    \n",
    "                      'example_mutDNA_2.csv',    \n",
    "                      'example_mutvirus_1.csv',\n",
    "                      'example_mutvirus_2.csv']\n",
    "\n",
    "## `replicates` stores which replicate each file in the list above corresponds to\n",
    "replicates = [1,  # replicate number for 'example_mutDNA_1.csv'\n",
    "              2,  # replicate number for 'example_mutDNA_2.csv'\n",
    "              1,  # replicate number for 'example_mutvirus_1.csv'\n",
    "              2]  # replicate number for 'example_mutvirus_2.csv'\n",
    "\n",
    "## `times` stores the collection time for each codon counts file\n",
    "## here we set the counts before selection as generation 0 and the counts after selection as generation 1\n",
    "## only differences in time are important -- setting the generations before and after selection as 1 and 2\n",
    "## would yield exactly the same results, for example\n",
    "times = [0,  # sample generation for 'example_mutDNA_1.csv'\n",
    "         0,  # sample generation for 'example_mutDNA_2.csv'\n",
    "         1,  # sample generation for 'example_mutvirus_1.csv'\n",
    "         1]  # sample generation for 'example_mutvirus_2.csv'\n",
    "\n",
    "## `freq_dir` specifies the directory where frequencies should be saved\n",
    "## to save in the current directory, use '.' or leave this variable out when calling the function\n",
    "freq_dir = '.'\n",
    "\n",
    "## `output_dir` specifies the directory where selection coefficients should be saved\n",
    "## to save in the current directory, use '.' or leave this variable out when calling the function\n",
    "output_dir = '.'\n",
    "\n",
    "## `plot_gamma` specifies whether or not to show a diagnostic plot of the correlation between\n",
    "## replicates as a function of the regularization strength `gamma` (set to True by default)\n",
    "plot_gamma = True\n",
    "\n",
    "\n",
    "#___ running popDMS\n",
    "\n",
    "# COMPUTE FREQUENCIES -- for codon counts, this should be very fast\n",
    "popDMS.compute_variant_frequencies_Bloom(name = name, \n",
    "                                         codon_counts_files = codon_counts_files,\n",
    "                                         replicates = replicates,\n",
    "                                         freq_dir = freq_dir)\n",
    "\n",
    "\n",
    "# INFER SELECTION COEFFICIENTS -- this should also be very fast for codon counts\n",
    "popDMS.infer_independent(name = name, \n",
    "                         n_replicates = n_replicates, \n",
    "                         corr_cutoff_pct = corr_cutoff_pct, \n",
    "                         freq_dir = freq_dir, \n",
    "                         output_dir = output_dir, \n",
    "                         plot_gamma = plot_gamma)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example using sequence counts\n",
    "\n",
    "Let's assume that we have a DMS data set that we refer to as `example`. In this data set, there were two experimental replicates. Sequence counts in [MAVE-HGVS format](https://www.mavedb.org/docs/mavehgvs/) were saved in the file `example_nucleotide_counts.csv`, which should be readable by `pandas`. Here we'll assume that the sequence counts before selection for the two replicates were stored in columns `replicate_A_0` and `replicate_B_0`, respectively. We'll assume that selection counts after selection, which we take to be 10 generations of replication, are stored in the columns `replicate_A_10` and `replicate_B_10`, respectively, in the same file. In addition to this data, we'll also need to save the nucleotide reference sequence in plain text, which here we assume is in a file `example_reference_sequence.dat`.\n",
    "\n",
    "The code block below would analyze this data set and store the inferred selection coefficients in a compressed CSV file `example_selection_coefficients.csv.gz`. This file can either be analyzed by `pandas` or similar packages while compressed, or it can be uncompressed to be viewed directly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import popDMS\n",
    "\n",
    "\n",
    "#___ popDMS parameters\n",
    "\n",
    "## `corr_cutoff_pct` sets the maximum fraction of drop in correlation per log10(gamma) for setting optimal regularization strength \n",
    "## 0.1 is a good default value\n",
    "corr_cutoff_pct = 0.1  \n",
    "\n",
    "\n",
    "#___ data parameters\n",
    "\n",
    "## `name` gives the name of the data set, which will be prepended to the files that popDMS saves\n",
    "name = 'example'\n",
    "\n",
    "## `n_replicates` specifies the number of experimental replicates for this data set\n",
    "n_replicates = 2\n",
    "\n",
    "## `reference_sequence_file` specifies the path to the reference sequence\n",
    "reference_sequence_file = 'example_reference_sequence.dat'\n",
    "\n",
    "## `haplotype_counts_file` gives the path to the file that stores the sequence counts\n",
    "haplotype_counts_file = 'example_nucleotide_counts.csv'\n",
    "\n",
    "## `time_points` specifies the times (in generations) at which data was collected\n",
    "## here we'll set the initial time as generation 0, and the second collection time after selection\n",
    "## as generation 10, but note that only the differences in time are important\n",
    "## setting these values to 10 and 20, respectively, would yield identical results\n",
    "time_points = [0, 10]\n",
    "\n",
    "## `time_cols` specifies the columns that store sequence counts sampled at the time above for each replicate\n",
    "time_cols = {1: ['replicate_A_0', 'replicate_A_10'],  # for the 1st replicate, counts stored in columns 'replicate_A_0'/'replicate_A_10' before/after selection\n",
    "             2: ['replicate_B_0', 'replicate_B_10']}  # for the 2nd replicate, counts stored in columns 'replicate_B_0'/'replicate_B_10' before/after selection\n",
    "\n",
    "## `freq_dir` specifies the directory where frequencies should be saved\n",
    "## to save in the current directory, use '.' or leave this variable out when calling the function\n",
    "freq_dir = '.'\n",
    "\n",
    "## `output_dir` specifies the directory where selection coefficients should be saved\n",
    "## to save in the current directory, use '.' or leave this variable out when calling the function\n",
    "output_dir = '.'\n",
    "\n",
    "## `plot_gamma` specifies whether or not to show a diagnostic plot of the correlation between\n",
    "## replicates as a function of the regularization strength `gamma` (set to True by default)\n",
    "plot_gamma = True\n",
    "\n",
    "\n",
    "#___ running popDMS\n",
    "\n",
    "# COMPUTE FREQUENCIES -- for sequence counts, this can be time-consuming depending on the size of the sequences,\n",
    "# the number of replicates, sequencing depth, and the number of time points included in the analysis\n",
    "# see examples at https://github.com/bartonlab/paper-DMS-inference for an idea of typical times\n",
    "popDMS.compute_variant_frequencies(name = name, \n",
    "                                   reference_sequence_file = reference_sequence_file, \n",
    "                                   haplotype_counts_file = haplotype_counts_file, \n",
    "                                   n_replicates = n_replicates, \n",
    "                                   time_points = time_points, \n",
    "                                   time_cols = time_cols, \n",
    "                                   freq_dir = freq_dir)\n",
    "\n",
    "\n",
    "# INFER SELECTION COEFFICIENTS -- typically, this step is relatively faster than computing frequencies above\n",
    "popDMS.infer_correlated(name = name, \n",
    "                        n_replicates = n_replicates, \n",
    "                        corr_cutoff_pct = corr_cutoff_pct, \n",
    "                        freq_dir = freq_dir, \n",
    "                        output_dir = output_dir, \n",
    "                        plot_gamma = plot_gamma)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example visualization of inferred selection coefficients\n",
    "\n",
    "Regardless of the data source (codon counts or sequence counts), selection coefficients inferred by popDMS will be saved in the same format. The code block below gives a brief example of calling the built-in visualization of inferred selection coefficients, assuming that this data is stored in a file `example_selection_coefficients.csv.gz`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import popDMS\n",
    "\n",
    "\n",
    "#___ popDMS parameters\n",
    "\n",
    "## `pop_file` specifies the path to the selection coefficients file\n",
    "pop_file = 'example_selection_coefficients.csv.gz'\n",
    "\n",
    "## `fig_file` specifies where to save the figure\n",
    "fig_file = 'example_plot.pdf'\n",
    "\n",
    "\n",
    "#___ running popDMS\n",
    "\n",
    "# PLOT SELECTION COEFFICIENT HEATMAP\n",
    "popDMS.fig_dms(pop_file = pop_file, \n",
    "               fig_file = fig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
